### PR TITLE
Add Delete Function to Vector Store Drivers

### PR DESCRIPTION
**Pull request by bjorger**
Partially implements #530 

Add Delete Function to All Classes That Inherit from the Base Vectorstore Class
Add Delete Function to Base Vectorstore Class
Implement Delete Function for MongoDB Vectorstore
Implement Delete Collection Function for MongoDB Vectorstore
Add Test for Delete Function for MongoDB Vectorstore

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -37,7 +37,10 @@ mail-parser = "^3.15.0"
 requests = "^2"
 lxml = "^4.9.3" # TODO: remove when https://github.com/adbar/trafilatura/issues/449 is fully resolved
 pillow = "^10.1.0"
-
+pytest = "^7.4.3"
+mongomock = "^4.1.2"
+pyright = "^1.1.343"
+google = "^3.0.0"
```
Comment (collindutter) [inline pyproject.toml:43]: Were these changes intentional? Some of them look like they belong in the `test` poetry group.

File: griptape/drivers/vector/base_vector_store_driver.py [hunk 2]
```diff
@@ -72,6 +72,9 @@ def upsert_text(
             **kwargs
         )
 
+    def delete(self):
+        raise NotImplementedError("This Method is not implemented yet")
```
Comment (collindutter) [inline griptape/drivers/vector/base_vector_store_driver.py:76]: Let's update the message to match other `NotImplementedError`s in the codebase. Something like:
`raise NotImplementedError(f"{self.__class__.__name__} does not support deletion.")`

**Review by collindutter — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (bjorger) [reply pyproject.toml:43]: Pytest, mongomock and pyright should not be there, you are correct!

But for google, I received an error in the testing pipeline for python 3.11 (or 3.12) that the dependency was missing, so I don't know if this should be in the testing dependencies or in the normal ones!

File: griptape/drivers/vector/base_vector_store_driver.py [hunk 3]
```diff
@@ -72,6 +72,10 @@ def upsert_text(
             **kwargs
         )
 
+    @abstractmethod
+    def delete(self, vector_id: str):
```
Comment (collindutter) [inline griptape/drivers/vector/base_vector_store_driver.py:76]: Let's rename to `delete_vector` for consistency with the other methods.

File: pyproject.toml [hunk 4]
```diff
@@ -37,6 +37,7 @@ mail-parser = "^3.15.0"
 requests = "^2"
 lxml = "^4.9.3" # TODO: remove when https://github.com/adbar/trafilatura/issues/449 is fully resolved
 pillow = "^10.1.0"
+google = "^3.0.0"
```
Comment (collindutter) [inline pyproject.toml:40]: Is this change intentional?

File: pyproject.toml [hunk 5]
```diff
@@ -106,6 +107,8 @@ optional = true
 pytest = "~=7.1"
 pytest-mock = "*"
 mongomock = "*"
+pyright = "^1.1.343"
+
```
Comment (collindutter) [inline pyproject.toml:111]: Is this change intentional?

**Review by collindutter — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 4 above)
Comment (bjorger) [reply pyproject.toml:40]: As stated above, without this dependency the python 3.11 pipeline failed for me - but I will remove the dependency and see if the pipeline runs successfully

Comment (collindutter) [reply pyproject.toml]: Ah sorry missed the previous review comment! Yeah let's try again without it, I can help troubleshoot if it still gives issues.

**Comment by collindutter** [comment]
@bjorger FYI don't worry about the `type-check` action.

File: griptape/drivers/vector/base_vector_store_driver.py [hunk 6]
```diff
@@ -72,6 +72,10 @@ def upsert_text(
             **kwargs
         )
 
+    @abstractmethod
+    def delete_vector(self, vector_id: str):
```
Comment (collindutter) [inline griptape/drivers/vector/base_vector_store_driver.py:76]: We should add an explicit return type of `None`.

File: pyproject.toml [hunk 7]
```diff
@@ -107,6 +107,7 @@ pytest = "~=7.1"
 pytest-mock = "*"
 mongomock = "*"
 
+
```
Comment (collindutter) [inline pyproject.toml:110]: Extra line, let's remove to leave `pyproject.toml` out of the changed files.

**Review by collindutter — CHANGES_REQUESTED** [review]

**Review by collindutter — DISMISSED** [review]

File: griptape/drivers/vector/mongodb_vector_store_driver.py [hunk 8]
```diff
@@ -135,3 +135,12 @@ def query(
...
         return results
+
+    def delete_vector(self, vector_id: str):
+        """Deletes the vector from the collection."""
+        collection = self.get_collection()
+        collection.delete_one({"_id": vector_id})
+
+    def delete_collection(self):
+        """Deletes the current collection from the database."""
+        self.get_collection().drop()
```
Comment (vasinov) [inline griptape/drivers/vector/mongodb_vector_store_driver.py:146]: Should the collection really be deleted here? If so, let's not move it into a separate method since it's a one-liner not used anywhere else.

**Review by vasinov — DISMISSED** [review]

File: griptape/drivers/vector/mongodb_vector_store_driver.py (hunk 8 above)
Comment (bjorger) [reply griptape/drivers/vector/mongodb_vector_store_driver.py:146]: I don't fully understand.
First of all, yes, the objective of this function is to delete the collection.
But where should I move it?

Comment (vasinov) [reply griptape/drivers/vector/mongodb_vector_store_driver.py:146]: Sorry, I misread the code. I don't think we need this method at all. The purpose of the driver is to handle vectors inside of a vector DB, not the DB itself.

Comment (bjorger) [reply griptape/drivers/vector/mongodb_vector_store_driver.py:146]: I, for once, need that functionality with our current use case, where we create embeddings of large documents and store them in individual collections.
Also, check out the screenshot attached. When you upload a big document as an embedding, it makes a bunch of vectors for each embedding. So, this is where the 'delete_collection' function comes in handy
image

Comment (vasinov) [reply griptape/drivers/vector/mongodb_vector_store_driver.py:146]: Your use case makes perfect sense but we are trying to keep all drivers pretty generic.

In this particular scenario, the user could chose between any number of ways to store vectors (e.g., single collection, collection per artifact, collection per artifact grouping, etc.), which could result in different `delete_*` functionality. We are also, in general, trying to reduce the number of one-off methods in drivers that could result in inconsistent UX

In your use case, would it make more sense to just use the Mongo SDK to delete the collection instead of moving that functionality into Griptape?

Comment (bjorger) [reply griptape/drivers/vector/mongodb_vector_store_driver.py:146]: I understand the UX aspect, but Griptape also creates the collection, therefore, it should provide a functionality to delete them as well, without going the way of using pymongo or any other SDK. 

Also, maybe I misunderstood you there, but there is no "out of the box" solution from griptape to store the vectors in the fashion that you just described. 
Furthermore, as I said above, when creating an embedding from a large document, you cannot store the entire thing in one vector, there will be several vectors created for that embedding, therefore, the delete_vector function would probably result in some weird work arounds to properly delete the entire embedding.

But this is my opinion - if you guys agree that the delete_collection function is not of use for griptape, I will remove it from the PR.

**Review by bjorger — COMMENTED** [review]
Work in review

**Review by collindutter — APPROVED** [review]
